r/R/dplyr-distinct.R

```diff
@@ -0,0 +1,71 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# The following S3 methods are registered on load if dplyr is present
+
+distinct.arrow_dplyr_query <- function(.data, ..., .keep_all = FALSE) {
+
+  if (.keep_all == TRUE) {
+    # After ARROW-13767 is merged, we can implement this via e.g.
+    # iris %>% group_by(Species) %>% slice(1) %>% ungroup()
+    abort("`distinct()` with `keep_all = TRUE` argument not supported in Arrow")
+  }
+
+  distinct_groups <- ensure_named_exprs(quos(...))
+
+  # Get ordering to use when returning data
+  cols_in_order <- intersect(
+    unique(c(names(.data), names(distinct_groups))),
+    names(distinct_groups)
+  )
+
+  # Get grouping in the data to add back in later, as the call to summarize()
+  # will remove it
+  gv <- dplyr::group_vars(.data)
+
+  vars_to_group <- unique(c(
+    names(distinct_groups),
+    gv
+  ))
+
+  if (length(vars_to_group) == 0) {
+    return(.data)
+  }
+
+  # Call mutate in case any columns are expressions
+  .data <- dplyr::mutate(.data, ...)
+
+  # This works as distinct(data, x, y) == summarise(group_by(data, x, y))
+  .data <- dplyr::group_by(.data, !!!syms(vars_to_group))
+  .data <- dplyr::summarize(.data)
+
+  # Add back in any grouping which existed in the data previously
+  if (length(gv) > 0) {
+    .data$group_by_vars <- gv
+  }
+
+  # Select the columns to return in the correct order
```

Comment:
Why are columns out of order? group_by should get the order right for you

Reply:
The order of columns returned by `group_by()` and then `summarise()` are the order in which they're supplied to `group_by()`, whereas the order of columns returned from `distinct()` are the order in which they occur in the data, even if there's a `group_by()` call previously.
```
> tbl %>%
+       group_by(some_grouping, int) %>%
+       summarise() %>%
+       collect()
`summarise()` has grouped output by 'some_grouping'. You can override using the `.groups` argument.
# A tibble: 10 × 2
# Groups:   some_grouping [2]
   some_grouping   int
           <dbl> <int>
 1             1     1
 2             1     3
 3             1     5
 4             1     7
 5             1     9
 6             2     2
 7             2     6
 8             2     8
 9             2    10
10             2    NA
> tbl %>%
+       group_by(some_grouping, int) %>%
+       distinct(lgl) %>%
+       collect()
# A tibble: 10 × 3
# Groups:   some_grouping, int [10]
     int lgl   some_grouping
   <int> <lgl>         <dbl>
 1     1 TRUE              1
 2     2 NA                2
 3     3 TRUE              1
 4    NA FALSE             2
 5     5 TRUE              1
 6     6 NA                2
 7     7 NA                1
 8     8 FALSE             2
 9     9 FALSE             1
10    10 NA                2
```

Reply:
Got it, that's worthy of a comment in the code :)

Maybe you can sort the group_by_vars to match their order in the data before calling summarize, then the result from summarize should match?

Reply:
Thanks for the suggestion, I'll do this!

Reply:
Added comment.  

Am no longer using `group_by_vars` after realising that `group_by()` has a `.add` argument, which seems cleaner to take advantage of instead of redoing all of the grouping myself. Therefore, have kept in the ordering code.